### PR TITLE
mail: add IMAP-subfolder checking

### PR DESCRIPTION
**Pull request by L0ric0**
This allows the mail module to check for new mail in IMAP-subfolders and not only in INBOX. Useful when automatically sorting mail.

### CODE REVIEW

**Comment by ultrabug** [comment]
Thanks a lot @L0ric0 

Before I merge this, could I ask for your opinion on #1526 maybe?

**Comment by lasers** [comment]
I want to look at this closely to understand the issue better. Maybe this
config is avoidable in first place in favor of placeholders or something
else (ie making it a default behaviour).

If it is pretty much same, then  it's possible that we could just loop it
instead to ensure everybody are happy here. I was not sure if looping was
necessary because I saw that I got the same value without...

It looks like looping is used in imap too along with other configs that
users might not need. Maybe I just need to understand what it was that I
missed out on... as this (and other things) was not included to keep
everything simple.

**Comment by L0ric0** [comment]
I would not set the scanning of sub-folders to the default behavior as the scanning can take a long time depending on your network connection and the amount of folders you have.

**Comment by tobes** [comment]
I think we should look at simplifying this module so that we can get rid of the need for a `check_imap_subfolders` option.

We know the format - this should tell us everything we need to know.  Maybe we need to rethink some of the logic though.

I'd also be keen to have some clearer documentation for the module too as the current stuff is poor and confusing imho

But the feature is good

File: py3status/modules/mail.py [hunk 1]
```diff
@@ -195,9 +198,20 @@ def mail(self):
...
-                    inbox.select(readonly=True)
-                    imap_data = inbox.search(None, '(UNSEEN)')
-                    count_mail = len(imap_data[1][0].split())
+                    if self.check_imap_subfolders:
+                        count_mail = 0
+                        for mbox in inbox.list()[1]:
+                            # Don't remove the quotes from the return string as
+                            # that would make the script unable to access all
+                            # mailfolders with special characters.
+                            mbox = mbox.decode('utf-8').split(' "." ')[1]
```
[the remark this answers is not in the document]
Comment (lasers) [reply py3status/modules/mail.py]: I got `"/"` directories here. `b'(\\HasNoChildren) "/" "INBOX"'`

[the remark this answers is not in the document]
Comment (L0ric0) [reply py3status/modules/mail.py]: ok then the second field gives us the folder separator and the fields are space separated. Is there a possibility to split only at spaces that are not in "? If not I would split at `' "." '` and `' "/" '` as for imap there shouldn't be any other folder separators and the strings should be unique enough to not show up in folder names.

[the remark this answers is not in the document]
Comment (lasers) [reply py3status/modules/mail.py]: What about:
```
from csv import reader
print([x[-1][:-1] for x in reader(out.splitlines(), delimiter=' ')])
```

**Comment by lasers** [comment]
I don't like creating three `imap` options because most of them won't be used. I would like to condense all three into a single parameter `filter` to keep things simple. My `mailfilter` does something like this, but is mostly untested. I could use eyes/opinions on this.

I think my `mailfilter` branch uses (whitelisted) `fnmatch` and is likely more than enough because of default `INBOX`, then users can add additional filter to match against. I'm unfamiliar on filtering... I mean, is it possible to whitelist and blacklist using `fnmatch` or do we need to create two separate `fnmatch` methods?

It's an issue of not knowing if we can whitelist and/or blacklist things with `fnmatch` (single option).

* To subscribe (default) `INBOX`, use `filters = ['INBOX']`.
* To subscribe more, use `filters = ['INBOX', '*Bank*']`.

If he want to subscribe all, but ignore SPAM folder... Can we do that with `fnmatch`... ie...
* `filters = ['*', '!*SPAM*']`

**Comment by L0ric0** [comment]
sorry for not replying so long. 
@lasers the filter `['*', '!*SPAM*']` dose not work as `fnmatch('*', 'SPAM')` will return true and `SPAM` will be added to the subscribed folders which is like putting an or between the filters.
Also I don't think that fnmatch is strong enough for this purpose as I haven't found a filter that will return false for only a word. re would probably be better as one can "easily" define a filter that matches all but spam as `^((?!Spam).)*$`

**Comment by lasers** [comment]
@L0ric0 Hi. I added `re.match` instead of `fnmatch` in same branch [mailfilter](https://github.com/ultrabug/py3status/compare/mailfilter). I'm not versed at all with `re` stuffs. Can you see if this does what you're proposing?

**Comment by L0ric0** [comment]
@lasers yes it is working, at last with the simple filter to remove the spam folder.

As the filters are working at the moment they are connected with an or. I would add this to the config as to not frustrate users.

Also the filters are casesensitive which is ok in my opinion but should also be mentiond.

Should I close this PR or pull from your branch?

**Comment by lasers** [comment]
I can push it onto your PR. Can you give few examples too? Eg, how to subscribe to multiple folders... and how to exclude a folder, how to filter case insensitive, etc. I will add them with the commit.

**Comment by L0ric0** [comment]
- include folder that contains pattern but no others: `pattern`
- include folders with caseinsensitive starting char: `[Pp]attern`
- include all folders: `.*`
- match pattern at begining of name: `^pattern`
- match pattern at end of name: `pattern$`
- exclude folders with spam in name and subscribe to all others: `^((?![Ss][Pp][Aa][Mm]).)*$` (every possible case
 
for more documentation see https://docs.python.org/3/library/re.html and any regex builder in the web

**Comment by lasers** [comment]
@L0ric0 Too much imo. I think users may be content with your filter suggestions. I'm not sure if we should even provide a list... or if we should stick with a single string filter? List is offered so users can "copy and paste" folders (from the log).

@ultrabug What do you want to do here? Destroy my commits here and close... to preserve @L0ric0's original code and open a new PR? We signed off on this (via `mailfilter` branch). I'd like to know how you want to proceed here with this PR.

**Comment by L0ric0** [comment]
@lasers then I would take out the list of examples but leave the notice that one can use regexes to filter.

**Comment by lasers** [comment]
Original commits preserved. Closing in favor of the other branch.

**Pull request closed without merging**